Add PodSecurityPolicy information to audit logs

CaoShuFeng opened the pull request:
Depends on: https://github.com/kubernetes/kubernetes/pull/58806
Fix #56209

**Release note**:
```release-note
PodSecurityPolicy admission information is added to audit logs
```

CaoShuFeng commented:
/assign @sttts @liggitt 
What's your opinion about this version?

sttts commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -153,11 +157,13 @@ func (c *PodSecurityPolicyPlugin) Validate(a admission.Attributes) error {
 	}
 	if apiequality.Semantic.DeepEqual(pod, allowedPod) {
+		ae := a.GetAuditEvent()
+		audit.LogAnnotations(ae, "podsecuritypolicy/validate", pspName)
```
we need better keys here, maybe `podsecuritypolicy.admission.k8s.io/policy`

soltysh replied:
👍 we should always use full names

sttts commented on plugin/pkg/auth/authorizer/rbac/rbac.go:
```diff
@@ -54,13 +54,21 @@ type RBACAuthorizer struct {
-	allowed bool
-	errors  []error
+	allowed            bool
+	clusterRoleBinding string
```
would move these into its own struct. maybe `struct reason`

sttts commented on plugin/pkg/auth/authorizer/rbac/rbac.go:
```diff
@@ -74,6 +82,21 @@ func (r *RBACAuthorizer) Authorize(requestAttributes authorizer.Attributes) (aut
 	if ruleCheckingVisitor.allowed {
+		ae := requestAttributes.GetAuditEvent()
+		if ae != nil {
+			if ruleCheckingVisitor.clusterRoleBinding != "" {
```
`len(...) != 0` is our convention

sttts commented on staging/src/k8s.io/apiserver/pkg/apis/audit/types.go:
```diff
@@ -125,6 +125,12 @@ type Event struct {
+	// PSP will store admission information in Annotations, like name of the policy which admits
+	// the request.
+	// +optional
+	Annotations map[string]string
```
@tallclair @crassirostris @ericchiang @soltysh this is about the API. Are we happy with this representation?

crassirostris replied:
I'm fine with this interface, but I remember someone was not happy with it

sttts replied:
We have plugin names (those which are also used in the apiserver flags). We could make this more structured using them if we concentrate on admission. We could render the whole chain here in the event.

RBAC would then be another step, probably with an explicit authorization field.

soltysh replied:
I just spoke with @sttts and we agreed it would be nice to:
1. be able to decide which admission and rbac plugins are logged, so that would require additional rule changes
2. Authorization logging happens in `authorizingVisitor` which is perfect, but admission happens only in PSP, I'd imagine this should happen a level higher to be able to do what I said in 1. 

@tallclair @crassirostris @ericchiang @CaoShuFeng opinions?

tallclair replied:
> be able to decide which admission and rbac plugins are logged, so that would require additional rule changes

This feels like something that should be resolved with filters, if we ever get around to implementing them. We don't currently allow fine grained selection of what's logged, so I think the current approach is consistent.

> Authorization logging happens in authorizingVisitor which is perfect, 

Only for RBAC, it's an implementation detail.

> but admission happens only in PSP, I'd imagine this should happen a level higher to be able to do what I said in 1.

I see this field as arbitrary metadata that plugins invoked in the serving path can add to. I think the authorization & admission interfaces should enable this on a case-by-case basis.

I'm happy with the current unstructured approach.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -94,7 +94,7 @@ func createHandler(r rest.NamedCreater, scope RequestScope, typer runtime.Object
 
 		userInfo, _ := request.UserFrom(ctx)
-		admissionAttributes := admission.NewAttributesRecord(obj, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Create, userInfo)
+		admissionAttributes := admission.NewAttributesRecordWithAudit(obj, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Create, userInfo, ae)
```
do we really need this new func name?

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -32,9 +33,10 @@ type attributesRecord struct {
 }
 
-func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info) Attributes {
+func NewAttributesRecordWithAudit(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info, ae *audit.Event) Attributes {
```
I would keep the old func name and make that argument mandatory (possibly nil)

CaoShuFeng replied:
OK.
I will modify all `NewAttributesRecord` functions.
```
$ LANG=C grep "NewAttributesRecord" * -r  | grep "\.go"  | wc -l
262
```
:)

sttts replied:
> 262

https://www.gnu.org/software/sed/manual/html_node/index.html ;)

sttts replied:
For easier review, please do that in another commit.

99% of those calls are in tests btw. Was surprised that the number is that high.

sttts commented:
> What's your opinion about this version?

Much better. It's what I expected :+1:

CaoShuFeng commented:
> Much better. It's what I expected 👍

Thanks. I will address all your comments and add some unit tests.

liggitt commented on pkg/registry/rbac/validation/rule.go:
```diff
@@ -42,7 +42,7 @@ type AuthorizationRuleResolver interface {
 	// VisitRulesFor invokes visitor() with each rule that applies to a given user in a given namespace, and each error encountered resolving those rules.
 	// If visitor() returns false, visiting is short-circuited.
-	VisitRulesFor(user user.Info, namespace string, visitor func(rule *rbac.PolicyRule, err error) bool)
+	VisitRulesFor(user user.Info, namespace string, visitor func(rule *rbac.PolicyRule, clusterRolebinding, clusterRole, roleBinding, role string, err error) bool)
```
prefer `(*rbac.RoleRef, *rbac.PolicyRule, ...)`, since we have the reference already?

liggitt replied:
hmm, I guess you need the binding name as well

liggitt commented on staging/src/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go:
```diff
@@ -61,6 +62,9 @@ type Attributes interface {
 	GetPath() string
+
+	// GetAuditEvent returns the audit event of the request
+	GetAuditEvent() *audit.Event
```
this interface is supposed to be fully serializable so it can be delegated via a subject access review. an audit.Event pointer breaks that

sttts replied:
Which means that we have to expose the audit annotations (or any other data structure we choose) for authz/n in a different way here. Also a delgated authorizer should fill in values.

liggitt replied:
authz already has the ability to return a reason along with the authorization decision. can we use that?

sttts replied:
@tallclair @crassirostris can you comment whether the reason is enough?

liggitt replied:
I would expect the following behavior:
* when authorization failed, the reason would contain aggregated reasons from all authorizers that provided reasons
* when authorization succeeded, the reason would contain the reason from the succeeding authorizer if it provided one

liggitt replied:
the authorizer reason is currently plumbed all the way up to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go#L49 which seems like a much easier place to get a handle to the audit event

liggitt replied:
strawman PR that makes RBAC return a detailed reason for what it allows that includes the subject, binding, and role: https://github.com/kubernetes/kubernetes/pull/58531

CaoShuFeng replied:
Done.

liggitt commented on plugin/pkg/auth/authorizer/rbac/rbac.go:
```diff
@@ -74,6 +82,21 @@ func (r *RBACAuthorizer) Authorize(requestAttributes authorizer.Attributes) (aut
 
 	r.authorizationRuleResolver.VisitRulesFor(requestAttributes.GetUser(), requestAttributes.GetNamespace(), ruleCheckingVisitor.visit)
 	if ruleCheckingVisitor.allowed {
+		ae := requestAttributes.GetAuditEvent()
```
getting an object out of requestAttributes in order to write back to it is a fundamental change in the way the authorizer attributes are used, and breaks remote delegation cases for that interface

sttts replied:
See above. We need an explicit data structure here. Not a huge event with unclear ownership.

CaoShuFeng commented on plugin/pkg/auth/authorizer/rbac/rbac.go:
```diff
@@ -62,7 +62,7 @@ type authorizingVisitor struct {
 	if rule != nil && RuleAllows(v.requestAttributes, rule) {
 		v.allowed = true
-		v.reason = fmt.Sprintf("allowed by %s", source.String())
+		v.reason = fmt.Sprintf("RBAC: allowed by %s", source.String())
```
`RBAC:`  is added to indicate that the reason comes from rbac.

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +50,8 @@ type Attributes interface {
 	// GetUserInfo is information about the requesting user
 	GetUserInfo() user.Info
+	// GetAuditEvent returns audit event of the request
+	GetAuditEvent() *audit.Event
```
I think a much more scoped change to this interface is needed (not exposing the full audit event)

The same comments about delegation, serialization, and ownership of the object apply here. Every attribute in `admission.Attributes` must be able to be serialized and sent to a mutating webhook admission plugin. How would an audit.Event be sent remotely?

The things an in-tree admission plugin can do should also be able to be done by a mutating webhook admission plugin. How would a remote admission webhook contribute audit annotations?

CaoShuFeng replied:
Thanks for the suggestion, I will investigate it.

tallclair replied:
Would it make sense to just add a `map[string]string` return value to the admit interface? Ditto for authorization & authentication interfaces?

CaoShuFeng replied:
I think the question is how to expeand this interface gracefully:

https://github.com/kubernetes/kubernetes/blob/1789a8513c745f274ff8ddc3c46948a4a2ff2468/staging/src/k8s.io/api/admission/v1beta1/types.go#L40

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +49,8 @@ type Attributes interface {
 	// GetUserInfo is information about the requesting user
 	GetUserInfo() user.Info
+	// GetAnnotations returns an annotation map for the admission chain, both ValidationInterface and MutationInterface are allowed to insert new Annotations.
+	GetAnnotations() map[string]string
```
@liggitt @sttts 
Please take a look.

My question:
Why not `SetAnnotations(key, value string)`?
If we use SetAnnotations:
1. annotations set by one admission controller will not exposed to each other. (What if an admission needs to read annotations from others?)
2. we can forbid one admission controller to modify annotations set by others.

GetAnnotations() is consistent with GetObject().

sttts replied:
> (What if an admission needs to read annotations from others?)

It shouldn't.

> we can forbid one admission controller to modify annotations set by others.

We trust them anyway.

I am not a fan of a completely free form string map. We should at least define a convention for the keys or even enforce that, e.g. `admission.k8s.io/PodSecurityPolicy-<key>: <value>` and `admission.k8s.io/MutatingWebhook-your.webhook.company.com-<key>: <value>`

CaoShuFeng replied:
I add SetAnnotations() to enforce the annotion format.

CaoShuFeng replied:
We can remove GetAnnotations(). If so, users can only set Annotation with function SetAnnotations(). And we need to change the signature of fucntion:
https://github.com/kubernetes/kubernetes/pull/58143/commits/3e253dc5f5f432c6f111d47fb299a7281cd440fd#diff-ba2d46d03943b63d3c3add85abff408bR157

sttts replied:
I would like to avoid opening this channel as it makes reordering of plugins hard later. So better remove it.

CaoShuFeng replied:
Done.

CaoShuFeng commented:
/test pull-kubernetes-bazel-test

liggitt commented:
>> (What if an admission needs to read annotations from others?)
>
> It shouldn't.

Agree. We don't want this to become a communication channel between admission plugins. We should retain the ability to reorder or run validating plugins in parallel

tallclair commented:
@CaoShuFeng - Can you break this PR up? It will be much easier to review. I suggest 1 PR to make the API changes (no need to populate the API - just add it + the generated files), and a separate PR to populate it.

CaoShuFeng commented:
> @CaoShuFeng - Can you break this PR up? It will be much easier to review. I suggest 1 PR to make the API changes (no need to populate the API - just add it + the generated files), and a separate PR to populate it.

Sure, will do it soon.

CaoShuFeng commented:
@tallclair 
Api changes to advanced audit:   https://github.com/kubernetes/kubernetes/pull/58806
add rbac info to advanced audit: https://github.com/kubernetes/kubernetes/pull/58807
add psp info to advanced audit : this pr
support annotations for admission webhook: https://github.com/kubernetes/kubernetes/pull/58679 (WIP)

tallclair commented:
Thanks for breaking this up, it's much easier to review now.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -32,9 +36,10 @@ type attributesRecord struct {
 }
 
-func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info) Attributes {
+func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info, annotations map[string]string) Attributes {
```
aren't the initial annotations always empty? Then we can get rid of all the trivial `nil` changes above.

CaoShuFeng replied:
Because we don't support GetAnnotations(). So we have to pass a annotation map to Admit/Validate functions. Otherwise we don't have a way to read it back.

```
annotations := make(map[string]string)
admissionAttributes := admission.NewAttributesRecord(obj, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Create, userInfo, annotations)
err = mutatingAdmission.Admit(admissionAttributes)
audit.LogAnnotations(ae, annotations)
```

sttts replied:
I was thinking we have a concrete Attributes object at that point with full access, but only pass the interface to the admission plugin. No need to hide a type on custruction.

tallclair replied:
Yes, I agree with @sttts here.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +49,13 @@ type Attributes interface {
+	// "podsecuritypolicy" is the name of the plugin, and admission.k8s.io is the name
+	// of the organization.
+	// Both ValidationInterface and MutationInterface are allowed to set Annotations.
+	SetAnnotations(plugin_name, key, value string) bool
```
pluginName or even only plugin. Underscore is against convention.

CaoShuFeng replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:
```diff
@@ -232,7 +232,7 @@ func (i *wrappedUpdatedObjectInfo) UpdatedObject(ctx genericapirequest.Context,
 
 // AdmissionToValidateObjectUpdateFunc converts validating admission to a rest validate object update func
-func AdmissionToValidateObjectUpdateFunc(admit admission.Interface, staticAttributes admission.Attributes) ValidateObjectUpdateFunc {
+func AdmissionToValidateObjectUpdateFunc(admit admission.Interface, staticAttributes admission.Attributes, annotations map[string]string) ValidateObjectUpdateFunc {
```
what are these annotations? As far as I see, the parameter is always nil.

CaoShuFeng replied:
Same with https://github.com/kubernetes/kubernetes/pull/58143#discussion_r165251614

sttts commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -135,6 +136,8 @@ func (c *PodSecurityPolicyPlugin) Admit(a admission.Attributes) error {
 		}
 		pod.ObjectMeta.Annotations[psputil.ValidatedPSPAnnotation] = pspName
+
+		a.SetAnnotations(AnnotationPluginName, "admit-policy", pspName)
```
wondering whether we really have to make the PSP plugin aware of its own name. The admission chain loop could add this logic of using the right qualified key.

CaoShuFeng replied:
Done.

sttts commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -43,7 +43,8 @@ import (
 
 const (
-	PluginName = "PodSecurityPolicy"
+	PluginName           = "PodSecurityPolicy"
```
we should also have the webhook as an example. My comment https://github.com/kubernetes/kubernetes/pull/58143/commits/e2b30326af2665f4e916e60571e57b5d06727972#r165015068 possibly conflicts with the webhook implementation. Only the webhook plugin (which is **one** plugin, or actually two for validating and mutating) knows about the registered webhooks. The admission chain does not.

hzxuzhonghu commented:
/cc

CaoShuFeng commented:
rebased

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +95,30 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+
+func (record *attributesRecord) SetAnnotations(pluginName string, annotations map[string]string) {
+	if record.annotations == nil {
+		return
```
Is it a conscious decision to silently discard annotation instead of creating a new map? Could you please link me to a discussion?

CaoShuFeng replied:
https://github.com/kubernetes/kubernetes/pull/58679#discussion_r168297676

crassirostris replied:
Why not always initialize this map to non-nil value?

CaoShuFeng replied:
> Why not always initialize this map to non-nil value?

The reason is explained here:
https://github.com/kubernetes/kubernetes/pull/58679#discussion_r168297676

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/attributes_test.go:
```diff
@@ -0,0 +1,83 @@
+	attr.SetAnnotation("INVALID-DNS-Subdomain", "policy", "privileged")
+	assert.Equal(t, annotations["INVALID-DNS-Subdomain/policy"], "", "invalid pluginName is not allowed ")
+
+	// test set concurrently
```
Even if it works, it's a pretty flacky test, isn't it?

CaoShuFeng replied:
Will this test make sort of sense, when run with `-race` option?

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -153,11 +155,12 @@ func (c *PodSecurityPolicyPlugin) Validate(a admission.Attributes) error {
 		return admission.NewForbidden(a, err)
 	}
 	if apiequality.Semantic.DeepEqual(pod, allowedPod) {
+		a.SetAnnotation(strings.ToLower(PluginName)+admission.AnnotationSuffix, "validate-policy", pspName)
```
Done.

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:
```diff
@@ -1825,11 +1826,27 @@ func testPSPAdmitAdvanced(testCaseName string, op kadmission.Operation, psps []*
 	}
 
 	err = plugin.Validate(attrs)
+	if shouldPassAdmit && op == kadmission.Create {
```
Done.

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-unit

tallclair commented:
I'm having a hard time tracing this change. I suggest avoiding modifying `NewAttributesRecord`.

CaoShuFeng committed:
support AddAnnotation in admission attributes

CaoShuFeng commented:
> I'm having a hard time tracing this change. I suggest avoiding modifying NewAttributesRecord.

I am working on this recently. New patches will come soon.
Thanks very much for your help.

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

CaoShuFeng commented:
This is ready for review.
Please take a look.

tallclair commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +49,17 @@ type Attributes interface {
+	// SetAnnotations sets annotations according to key-value pairs in annotations map. The keys of annotations map should be qualified, e.g., podsecuritypolicy.admission.k8s.io/admit-policy,
+	// where "podsecuritypolicy" is the name of the plugin, "admission.k8s.io" is the name of the organization, "admit-policy" is the key name.
+	// Both ValidationInterface and MutationInterface are allowed to set Annotations.
+	SetAnnotations(annotations map[string]string)
```
I'm still not sold that mutable attributes is the right interface for this. What if instead we modify the admission interfaces to return a result struct that (optionally) includes annotations?

```go
Admit(a Attributes) (Result, error)
Validate(a Attributes) (Result, error)

type Result struct {
  Decision Decision
  Reason string  // Do we care about this?
  AuditInfo map[string]string
  // Optional: HandlerName string
}

type Decision string
const (
  DecisionAccept Decision = "Accept"
  DecisionReject Decision = "Reject"
)
```

@sttts @liggitt WDYT?

CaoShuFeng replied:
Waiting for @sttts and @liggitt s' suggestions.

sttts replied:
Note that the new object returned via `GetObject()` is also mutable. 

Also note that the synchronization of those mutations is not implicit by the attributes implementation, but done by the callee. With the annotation funcs we differ from those semantics.

CaoShuFeng replied:
Updated the comment.

tallclair replied:
@sttts Are you suggesting that since the object is mutable it's reasonable to mutate the annotations on this interface?

One difference is that the object is only considered mutable during the mutating admission phase, while the annotations can also be modified during the validating phase.

sttts replied:
@tallclair was just an observation, am not suggesting that this is a good interface. In general, let's keep the interface small (only `AddAnnotation`, not getter). But at that point we are not far from your `Result` type which I would also like.

CaoShuFeng replied:
Thanks very for your suggestions will work on this soon.

CaoShuFeng replied:
> In general, let's keep the interface small (only AddAnnotation, not getter). 

Done.

@tallclair @sttts Please take a look.
This version only adds one function to the interface.

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -84,6 +84,7 @@ var _ admission.MutationInterface = &PodSecurityPolicyPlugin{}
 var _ admission.ValidationInterface = &PodSecurityPolicyPlugin{}
 var _ genericadmissioninit.WantsAuthorizer = &PodSecurityPolicyPlugin{}
 var _ kubeapiserveradmission.WantsInternalKubeInformerFactory = &PodSecurityPolicyPlugin{}
+var fullPluginName string = strings.ToLower(PluginName) + admission.AnnotationSuffix
```
done.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -32,19 +38,24 @@ type attributesRecord struct {
+		object:         object,
+		oldObject:      oldObject,
+		userInfo:       userInfo,
+		annotationLock: sync.RWMutex{},
```
done.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +94,45 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
 }
+
+func (record *attributesRecord) GetAnnotations() map[string]string {
+	if record.annotations == nil {
```
Done.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +94,45 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+}
+
+func (record *attributesRecord) SetAnnotations(annotations map[string]string) {
+	for key, value := range annotations {
```
Added a lock here.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -32,6 +38,10 @@ type attributesRecord struct {
+	annotations map[string]string
+	// other elements are always accessed in single goroutine.
+	// But ValidatingAdmissionWebhook sets annotations concurrently.
+	annotationLock sync.RWMutex
```
nit: but on-top of annotations and add empty line above

CaoShuFeng replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +93,54 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+func (record *attributesRecord) setAnnotation(key, value string, lock bool) {
+	parts := strings.Split(key, "/")
+	if len(parts) != 2 {
+		glog.Warningf("Annotation key has invalid format, the right format is a DNS subdomain prefix and '/' and key name. (e.g. 'podsecuritypolicy.admission.k8s.io/admit-policy')")
```
why is this not an error we pass down to the caller?

CaoShuFeng replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +49,17 @@ type Attributes interface {
 	GetUserInfo() user.Info
+	// GetAnnotations returns a copy of saved annotations. Modification to the returned copy of annotations will not be saved to Attributes. Use SetAnnotation/SetAnnotations
+	// to set annotations of Attributes.
+	GetAnnotations() map[string]string
```
do we really have to leak the getter here? At all call sites we have a concrete object, not the interface. We don't need that getter in the interface.

CaoShuFeng replied:
We can pass an annotation map to `NewAttributesRecord` to avoid this `GetAnnotations`.
In this case annotation would work like other elements. But it needs a lot of modification of codes.

Comments from @tallclair 
> I'm having a hard time tracing this change. I suggest avoiding modifying NewAttributesRecord.

hzxuzhonghu commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +93,54 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+	if record.annotations == nil {
+		return nil
+	}
+	cp := make(map[string]string)
```
make with length of record.annotations

CaoShuFeng replied:
Done.

hzxuzhonghu commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +93,54 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+	defer record.annotationLock.Unlock()
+
+	for key, value := range annotations {
+		record.setAnnotation(key, value, false)
```
not sure why can not set by `record.annotations = annotations`?

hzxuzhonghu replied:
ic, you do not mean to replace all annotations , just an add.

hzxuzhonghu commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +93,54 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+	record.setAnnotation(key, value, true)
+}
+
+func (record *attributesRecord) setAnnotation(key, value string, lock bool) {
```
when should not lock?

hzxuzhonghu commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +93,54 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+		record.annotations = make(map[string]string)
+	}
+	if v, ok := record.annotations[key]; ok && v != value {
+		glog.V(5).Infof("admission annotation overwrite, key:%q, old value: %q, new value:%q", key, record.annotations[key], value)
```
what's this for? Only for debug? Then should delete this if

CaoShuFeng replied:
We only log it when one key-value pair is overwroten.

CaoShuFeng committed:
add WithAudit admission decorator

WithAudit admission decorator log annotations to audit events set by
the decorated admission controller

CaoShuFeng commented:
/test pull-kubernetes-integration

CaoShuFeng commented:
/test pull-kubernetes-integration

CaoShuFeng committed:
Log policy name from pod security policy

CaoShuFeng commented:
kindly ping @sttts @tallclair for this.

CaoShuFeng commented:
rebased

sttts commented:
I am still not a fan of putting everything into annotations. But I also don't feel so strongly about it. @tallclair wdyt?

tallclair commented:
> I am still not a fan of putting everything into annotations. But I also don't feel so strongly about it. @tallclair wdyt?

@sttts At the moment, I think annotations is the best solution we have for dealing with all the different plugin points, since we can't bake them into first-class fields. We may be able to add some structure for some cases -  e.g.  `AuthorizationInfo { Decider string; Reason string; Extra map[string]string }`, but I think the annotations are a good starting point (maybe we should have called them "ExtraInfo" instead).

sttts commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -84,6 +84,7 @@ var _ admission.MutationInterface = &PodSecurityPolicyPlugin{}
 var _ admission.ValidationInterface = &PodSecurityPolicyPlugin{}
 var _ genericadmissioninit.WantsAuthorizer = &PodSecurityPolicyPlugin{}
 var _ kubeapiserveradmission.WantsInternalKubeInformerFactory = &PodSecurityPolicyPlugin{}
+var fullPluginName = strings.ToLower(PluginName) + admission.AnnotationSuffix
```
admissionPluginDomain. Can we standardize on some schema? `lowercase(pluginname).admission.k8s.io`?

sttts replied:
nvmd. It's exactly what the code is doing 👍

sttts replied:
just the var name can be improved.

CaoShuFeng replied:
done

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -32,6 +39,11 @@ type attributesRecord struct {
+	// other elements are always accessed in single goroutine.
+	// But ValidatingAdmissionWebhook add annotations concurrently.
+	annotations    map[string]string
+	annotationLock sync.RWMutex
```
nit: annotationsLock

CaoShuFeng replied:
done

hzxuzhonghu commented on staging/src/k8s.io/apiserver/pkg/admission/audit.go:
```diff
@@ -0,0 +1,70 @@
+
+// auditHandler logs annotations set by other admission handlers
+type auditHandler struct {
+	Interface
```
seems redundant

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -154,11 +158,14 @@ func (c *PodSecurityPolicyPlugin) Validate(a admission.Attributes) error {
 	}
 	if apiequality.Semantic.DeepEqual(pod, allowedPod) {
+		if err := a.AddAnnotation(fullPluginName+"/"+"validate-policy", pspName); err != nil {
+			glog.Errorf("failed to set annotation to attributes: %s", err.Error())
```
done

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +95,50 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+	return nil
+}
+
+func (record *attributesRecord) addAnnotation(key, value string) {
```
done

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/audit.go:
```diff
@@ -0,0 +1,70 @@
+	if mutator, ok := handler.Interface.(MutationInterface); ok {
+		err = mutator.Admit(a)
+	}
+	handler.logAnnotations(a)
```
done

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +95,47 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+func checkKeyFormat(key string) error {
+	parts := strings.Split(key, "/")
+	if len(parts) != 2 {
+		return fmt.Errorf("Annotation key has invalid format, the right format is a DNS subdomain prefix and '/' and key name. (e.g. 'podsecuritypolicy.admission.k8s.io/admit-policy')")
```
lower case, also below

CaoShuFeng replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +49,12 @@ type Attributes interface {
+	// AddAnnotation sets annotation according to key-value pair. The key should be qualified, e.g., podsecuritypolicy.admission.k8s.io/admit-policy, where
+	// "podsecuritypolicy" is the name of the plugin, "admission.k8s.io" is the name of the organization, "admit-policy" is the key name.
+	// An error is returned if the format of key is invalid.
+	// Both ValidationInterface and MutationInterface are allowed to add Annotations.
```
what happens on conflict?

CaoShuFeng replied:
If a annotation key is set more than once, it will be overwritten.

One key should always be set only be one admission controller, so conflict seldom happens.

The use case of overwritten behavior:
1. an admission made a decision and set a key.
2. later, it made a new decision and overwrite the key.

CaoShuFeng replied:
Added the behavior to comments.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -82,6 +88,8 @@ const (
 	Delete  Operation = "DELETE"
 	Connect Operation = "CONNECT"
+
+	AnnotationSuffix = ".admission.k8s.io"
```
AuditEventAnnotationSuffix

CaoShuFeng replied:
Done.

sttts commented:
Left some last nit comments. Overall I am fine with this. 

@tallclair @liggitt anything left from your side?

tallclair reviewed (commented):
Mostly nits, lets try to get this into 1.11. Sorry for the slow review turn around.

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission.go:
```diff
@@ -84,6 +84,7 @@ var _ admission.MutationInterface = &PodSecurityPolicyPlugin{}
 var _ admission.ValidationInterface = &PodSecurityPolicyPlugin{}
 var _ genericadmissioninit.WantsAuthorizer = &PodSecurityPolicyPlugin{}
 var _ kubeapiserveradmission.WantsInternalKubeInformerFactory = &PodSecurityPolicyPlugin{}
+var admissionPluginDomain = strings.ToLower(PluginName) + admission.AuditEventAnnotationSuffix
```
> nit: and rename the variable to auditKeyPrefix

Done.

Let's name this based on the API rather than the implementation

But not all admission controller has an API. What do you think?

tallclair replied:
Eh, let's deal with it on a case-by-case basis. I'd prefer the annotation reflect the externally facing behavior rather than the implementation.

CaoShuFeng replied:
Done.

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:
```diff
@@ -1798,11 +1798,24 @@ func testPSPAdmit(testCaseName string, psps []*policy.PodSecurityPolicy, pod *ka
+
+func (f fakeAttributes) AddAnnotation(k, v string) error {
+	f.annotations[k] = v
+	return nil
```
Done.

CaoShuFeng commented on plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:
```diff
@@ -1832,11 +1845,27 @@ func testPSPAdmitAdvanced(testCaseName string, op kadmission.Operation, psps []*
+	}
+}
+
+func validateAuditAnnotation(testCaseName string, annotations map[string]string, key, value string, t *testing.T) {
```
Done.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +95,47 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+		record.annotations = make(map[string]string)
+	}
+	if v, ok := record.annotations[key]; ok && v != value {
+		glog.V(5).Infof("admission annotation overwrite, key:%q, old value: %q, new value:%q", key, record.annotations[key], value)
```
Done.

sttts replied:
👍🏼

tallclair commented:
/status approved-for-milestone

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -49,6 +49,23 @@ type Attributes interface {
+// AnnotationsGetter allows users to get annotations from Attributes. An alternate Attribute should implement
+// this interface.
+type AnnotationsGetter interface {
+	GetAnnotations() map[string]string
```
why do we need this?

sttts replied:
I don't see this used. Premature abstraction? Would drop it.

CaoShuFeng replied:
This interface is used according to @liggitt 's suggestion.

https://github.com/kubernetes/kubernetes/commit/d7474b875587c32b97f0cd93553ad1a9c9aa8441#r29220311

CaoShuFeng replied:
It's used in WithAudit decorator to check the attribute interface.
Would be useful for alternate Attributes.

sttts replied:
ok, sounds reasonable.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -83,3 +93,48 @@ func (record *attributesRecord) GetOldObject() runtime.Object {
+		record.annotations = make(map[string]string)
+	}
+	if v, ok := record.annotations[key]; ok && v != value {
+		return fmt.Errorf("admission annotations are not allowd to be overwritten, key:%q, old value: %q, new value:%q", key, record.annotations[key], value)
```
Yes.
According to @tallclair 's suggestion.

sttts replied:
👍

sttts commented:
/lgtm
/approve

sttts commented:
@tallclair @liggitt approved?

liggitt commented:
/approve

jberkus commented:
/retest pull-kubernetes-e2e-kops-aws

liggitt commented:
/retest
